### ISSUE TITLE
High perplexity and FoSS adjustment in GPT-2 kNN-LM

### ISSUE THREAD

**Issue by lihe27341** [neulab/knn-transformers#17]
Hi @urialon !
Thank you for your excellent work and the resources provided in this repository.

I am currently trying to replicate the kNN-LM results for GPT-2 but have encountered an issue with the perplexity being higher than expected. Using the default hyperparameters, I obtained a PPL of 17.34 vs 12.57.

#### Tuning k, knn_temp, and lambda:
Adjusting these hyperparameters improved the PPL to some extent. For example, setting knn_temp to 50 resulted in a PPL of 14.84. Despite trying numerous combinations of these hyperparameters, I could not achieve the reported PPL of 12.57. Could you share the specific hyperparameter settings that were used to obtain this result?

#### Modifying the Fraction of Saved Searches (FoSS):
In the README, I noticed that changing the FoSS value significantly impacts kNN-LM performance. However, in traditional kNN-LM implementations, a full kNN search is conducted for every token, which seems to make the FoSS value fixed. I am unsure how to adjust the FoSS value effectively in this context. Could you provide any guidance or examples on how to modify it to optimize performance?
<img width="277" alt="微信图片_20241205002258" src="https://github.com/user-attachments/assets/3f06a789-05fd-41a4-8ce4-a5dd289880c9">
Thank you!

**Comment by urialon** [comment]
Hi,
Thank you for your interest in our work!

Did you try running with our datastore, our index etc? (Without tuning
anything)

**Comment by lihe27341** [comment]
Thank you for your reply!

Yes, I have tried running with both your datastore and index, and I am able to achieve the same PPL as yours with both the base model and RetoMaton.

**Comment by urialon** [comment]
So where is the problem?

**Comment by lihe27341** [comment]
The issue is with the GPT-2 kNN-LM's PPL. Without changing any hyperparameters, I get a PPL of 17.34 compared to your reported 12.57, which is significantly higher. (Although for the base model and RetoMaton, I do achieve results consistent with yours.)

**Comment by urialon** [comment]
But it is happening only when you create the datastore, right?
With my datastore it works fine?

**Comment by lihe27341** [comment]
No, that’s not the case. Whether I use your datastore or create my own, I get the same results: a PPL of 17.34 compared to your reported 12.57.

<img width="379" alt="微信图片_20241205092339" src="https://github.com/user-attachments/assets/53e2a6fa-d4bf-461c-8ffb-b9fbb27ae4f2">
